app/experimenter/experiments/tests/test_views.py

```diff
@@ -721,6 +732,66 @@ def test_view_renders_correctly(self):
         self.assertTemplateUsed(response, "experiments/detail_draft.html")
         self.assertTemplateUsed(response, "experiments/detail_base.html")
 
+    def test_view_renders_locales_correctly(self):
+        user_email = "user@example.com"
+        experiment = ExperimentFactory.create_with_status(
+            Experiment.STATUS_DRAFT
+        )
+        experiment.locales.add(LocaleFactory(code="yy", name="Why"))
+        experiment.locales.add(LocaleFactory(code="xx", name="Xess"))
+        response = self.client.get(
+            reverse("experiments-detail", kwargs={"slug": experiment.slug}),
+            **{settings.OPENIDC_EMAIL_HEADER: user_email},
+        )
+        self.assertEqual(response.status_code, 200)
+        html = response.content.decode(response.charset)
```

Comment:
I know you've mentioned you don't like end-to-end tests but I insist that it's valuable to test the template rendering still. In particular, during the development I had a typo of my own in the template and the tests passed. That's because I had written all the template changes, .py changes and tests *except* the tests that check the template. Then when I did one last manual testing in the browser I noticed my own bug. I had an extra stray bracket and not test noticed it. So that's when I went back to this file and added these tests that study the generated HTML. 

If you want, I can delete this and accept the risk in template code. Your call.

Reply:
I've been doing Django development long enough that I have not even able to find value in this kind of test.  For the case you're describing, either you would've found it through manual testing or QA would have.  I've been trying to keep up a habit of including screenshots in the PR for changes that include UI.  I'd suggest removing this test and including screenshots in the PR.

Reply:
But the templates are non-trivial. It contains code that can have bugs. Testing implementation details mean that you are dependent on details that don't matter for the greater good which is to reduce bad experiences for end users. Granted, we could extend that and switch to pure HTTP instead of Django's nifty testclient but that'd make it hard to mock breaking networks etc. 

These are looong posts and I don't want to distract by "forcing" you to read long texts but these are *some* well prosed arguments for why it's an anti-pattern to test implementation details. 

https://kentcdodds.com/blog/testing-implementation-details
https://softwareengineering.stackexchange.com/a/380296
https://stackoverflow.com/questions/856115/should-one-test-internal-implementation-or-only-test-public-behaviour

If you feel confident about this I can simply revert the test and try to remember that going forward for this project.

Reply:
I just removed the tests that checks the output.